googler

```diff
@@ -740,313 +1324,316 @@ class Result:
-        printerr('')
-
-    if output_format == 0: # Regular output
-        for r in results:
-            urlindex.update(r.print_entry())
-        if filtered:
-            printerr('** Enter "unfilter" to show similar results Google omitted.')
-    elif output_format == 1: # Json output
-        results_object = [r.json_object() for r in results]
-        print(json.dumps(results_object, indent=2, sort_keys=True, ensure_ascii=False))
-
-    return results, urlindex, filtered
-
-
-def read_next_command():
-    """Show omniprompt and read user command."""
-
-    global colorize, colors
-
-    message = 'googler (? for help)'
-    prompt = (colors.prompt + message + colors.reset + ' ') if colorize else (message + ': ')
-    try:
-        return input(prompt)
-    except EOFError:
-        quit_program(exitcode=0)
-
-
-def construct_search_url(baseurl, keywords, site=None):
-    """Construct the search URL with keywords and optionally a site.
-
-    baseurl should already contain queries such as start and num. This
-    function only adds the q query.
-
-    keywords should either be a list of keywords, or a single
-    space-delimited string of keywords.
+    Command line interpreter and executor class for googler.
+
+    Inspired by PSL cmd.Cmd.
+
+    Parameters
+    ----------
+    opts : argparse.Namespace
+        Options and/or arguments.
+
+    Attributes
+    ----------
+    options : argparse.Namespace
+        Options that are currently in effect. Read-only attribute.
+    keywords : str or list or strs
+        Current keywords. Read-only attribute
+
+    Methods
+    -------
+    fetch()
+    display_results(prelude='\n', json_output=False)
+    fetch_and_display(prelude='\n', json_output=False, interactive=True)
+    read_next_command()
+    help()
+    cmdloop()
     """
-    url = baseurl
-    if isinstance(keywords, list):
-        url += 'q=' + '+'.join([urllib.parse.quote_plus(kw) for kw in keywords])
-    else:
-        url += 'q=' + urllib.parse.quote_plus(keywords)
-    if site:
-        url += '+site:' + urllib.parse.quote_plus(site)
-    debugp('Search URL [%s : %s]' % (server, url))
-    return url
 
+    # Class variables
+    colors = None
 
-def open_url(url):
-    _stderr = os.dup(2)
-    os.close(2)
-    _stdout = os.dup(1)
-    os.close(1)
-    fd = os.open(os.devnull, os.O_RDWR)
-    os.dup2(fd, 2)
-    os.dup2(fd, 1)
-    try:
-        webbrowser.open(url)
-    finally:
-        os.close(fd)
-        os.dup2(_stderr, 2)
-        os.dup2(_stdout, 1)
-
+    def __init__(self, opts):
+        super().__init__()
 
-# Messaging wrappers
+        self._opts = opts
 
-def message(msg, level='INFO', ansi=None):
-    """Print message, led with [%(level)s] where level is a string, to
-    stderr.  If ansi is not None and the global variable colorize is True,
-    then print ansi at the beginning, and close with \x1b[0m.
-    """
+        self._google_url = GoogleUrl(opts)
```

Comment:
Per new query. If you don't change the keywords, it's just updated. If you change the keywords, a new instance is created with the given options. The reason is to get rid of ephemeral changes that shouldn't be kept between queries; currently, the only thing that belongs to that category is the `filter` query.

`GoogleUrl` is fast. There's not a single expensive operation in it.